### PR TITLE
Update S1764: Comparing the same object with Object.Equals() method

### LINKED ISSUE #235: Update S1764: Comparing the same object with Object.Equals() method
The goal of this ticket is to extend the scope of rule [RSPEC-1764](https://jira.sonarsource.com/browse/RSPEC-2739) "Identical expressions should not be used on both sides of a binary operator" to make it detect the following code patterns:

```
a.Equals(a);
Object.Equals(a,a);
```

### PR DESCRIPTION
**Pull request by michalb-sonar**
Fix #235

### CODE REVIEW

File: sonaranalyzer-dotnet/its/expected/akka.net/Akka.Persistence-{FCA84DEA-C118-424B-9EB8-34375DFEF18A}-S1764.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{
+"issues":  [
+{
+"id":  "S1764",
+"message":  "Correct one of the identical expressions on both sides of Object.Equals.",
+"location":  [
+{
+"uri":  "akka.net\src\core\Akka.Persistence\AtLeastOnceDeliverySemantic.cs",
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/its/expected/akka.net/Akka.Persistence-{FCA84DEA-C118-424B-9EB8-34375DFEF18A}-S1764.json]: Wow, that's unexpected! Here's the code:
```
return Equals(DeliveryId, other.DeliveryId)
           && Equals(UnconfirmedDeliveries, UnconfirmedDeliveries);
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/KnownMethods.cs [hunk 2]
```diff
@@ -42,14 +42,27 @@ public static bool IsMainMethod(this IMethodSymbol methodSymbol)
         public static bool IsObjectEquals(this IMethodSymbol methodSymbol)
         {
             return methodSymbol != null &&
-                methodSymbol.IsOverride &&
                 methodSymbol.MethodKind == MethodKind.Ordinary &&
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/KnownMethods.cs]: Let's add a or condition here (`IsOverride` or is in object).

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/BinaryOperationWithIdenticalExpressionsBase.cs [hunk 3]
```diff
@@ -25,6 +25,8 @@ namespace SonarAnalyzer.Rules
     public abstract class BinaryOperationWithIdenticalExpressionsBase : SonarDiagnosticAnalyzer
     {
         internal const string DiagnosticId = "S1764";
-        internal const string MessageFormat = "Identical sub-expressions on both sides of operator '{0}'.";
+
+        protected const string OperatorMessageFormat = "Correct one of the identical expressions on both sides of operator '{0}'.";
+        protected const string EqualsMessage = "Correct one of the identical expressions on both sides of Object.Equals.";
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/BinaryOperationWithIdenticalExpressionsBase.cs]: I would put the `Object.Equals` between single quotes.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/BinaryOperationWithIdenticalExpressions.cs [hunk 4]
```diff
@@ -40,6 +42,25 @@ public void Test(bool a, bool b)
...
+//                        ^ {{Correct one of the identical expressions on both sides of Object.Equals.}}
+//                           ^ Secondary@-1
+
+            var o = new object();
+
+            o.Equals(o);
+//          ^ {{Correct one of the identical expressions on both sides of Object.Equals.}}
+//                   ^ Secondary@-1
+
+            (new object()).Equals(new object());
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/BinaryOperationWithIdenticalExpressions.cs]: Ask Jesse about this case. Shall we report or not? The expression is the same but not the actual values. Although I agree that this is probably a fishy code.

Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/BinaryOperationWithIdenticalExpressions.cs]: We've discussed that with @jccollet and decided to report on that.

There may be corner cases where this is a false positve (like Equals(list.Next, list.Next)), but that would clearly be a clear code smell.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/BinaryOperationWithIdenticalExpressionsBase.cs [hunk 5]
```diff
@@ -25,6 +25,8 @@ namespace SonarAnalyzer.Rules
     public abstract class BinaryOperationWithIdenticalExpressionsBase : SonarDiagnosticAnalyzer
     {
         internal const string DiagnosticId = "S1764";
-        internal const string MessageFormat = "Identical sub-expressions on both sides of operator '{0}'.";
+
+        protected const string OperatorMessageFormat = "Correct one of the identical expressions on both sides of operator '{0}'.";
+        protected const string EqualsMessage = "Correct one of the identical expressions on both sides of 'Object.Equals'.";
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/BinaryOperationWithIdenticalExpressionsBase.cs]: this should probably be "... expressions in both arguments of ..." (if you change the message don't forget to update the unit tests and the integration test results)

Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/BinaryOperationWithIdenticalExpressionsBase.cs]: Also, I think we need to add a new ticket extend the scope of this rule and check for string.Equals(x, y), string.Equals(x, y, StringComparison) and perhaps other Equals(x, y) methods as well.
